WordPress/gutenberg
# Patterns: Refactor pattern categories to use existing endpoint to merge in user categories

## What?
Merges the user and theme pattern categories in the `/v2/block-patterns/categories` endpoint instead of merging them in multiple places in the UI

## Why?
Currently, the merging of these categories is duplicated in the block and site editor. It would be better to handle this in a single place, and it makes sense to do it on the API endpoint.

## How?
In order to maintain backwards compat with the existing selectors a new resolver and selector is added for `getAllPatternCategories` and the endpoint has been modified to only return the user categories and additional fields if explicitly specified in the request.

There is a corresponding PR to port the changes to the API endpoint to core [here](https://github.com/WordPress/wordpress-develop/pull/5621).

## Testing Instructions

- Add some patterns in the block editor, and assign some to existing theme/core pattern categories and some to new custom categories
- Make sure that the patterns appear under the theme/core categories in the inserter `Patterns` tab, and that the new custom categories also appear in the `Patterns` tab with the correct patterns listed
- When adding a new pattern make sure that the theme/core and custom categories appear in the category select list
- In the site editor make sure all the categories appear in the Patterns section, with the patterns listed under the correct categories
- Try adding a new pattern from the site editor `+` option at top of navigation and make sure new categories appear in the left category panel after adding
- Try renaming a custom category with overflow menu at top right of Patterns section when category selected, and make sure the change is reflected everywhere without having to refresh.
